ReactAndroid/src/main/java/com/facebook/react/modules/core/Timing.java

```diff
@@ -107,13 +120,84 @@ public void doFrame(long frameTimeNanos) {
+
+    @Override
+    public void doFrame(long frameTimeNanos) {
+      if (isPaused.get()) {
+        return;
+      }
+
+      // If the JS thread is busy for multiple frames we cancel any other pending runnable.
+      if (mCurrentIdleCallbackRunnable != null) {
+        mCurrentIdleCallbackRunnable.cancel();
+      }
+
+      mCurrentIdleCallbackRunnable = new IdleCallbackRunnable(frameTimeNanos);
+      getReactApplicationContext().runOnJSQueueThread(mCurrentIdleCallbackRunnable);
+
+      Assertions.assertNotNull(mReactChoreographer).postFrameCallback(
+          ReactChoreographer.CallbackType.IDLE_EVENT,
+          this);
+    }
+  }
+
+  private class IdleCallbackRunnable implements Runnable {
+    private volatile boolean mCancelled = false;
+    private final long mFrameStartTime;
+
+    public IdleCallbackRunnable(long frameStartTime) {
+      mFrameStartTime = frameStartTime;
+    }
+
+    @Override
+    public void run() {
+      if (mCancelled) {
+        return;
+      }
+
+      long frameTimeMillis = mFrameStartTime / 1000000;
+      long timeSinceBoot = SystemClock.uptimeMillis();
+      long frameTimeElapsed = timeSinceBoot - frameTimeMillis;
+      long time = SystemClock.currentTimeMillis();
+      long absoluteFrameStartTime = time - frameTimeElapsed;
+
+      if (FRAME_DURATION - (float)frameTimeElapsed < IDLE_CALLBACK_FRAME_DEADLINE) {
+        return;
+      }
+
+      mIdleCallbackContextsToCall.clear();
+      synchronized (mIdleCallbackGuard) {
+        mIdleCallbackContextsToCall.addAll(mSendIdleEventsExecutorTokens);
+      }
+
+      for (ExecutorToken context : mIdleCallbackContextsToCall) {
+        getReactApplicationContext().getJSModule(context, JSTimersExecution.class)
+            .callIdleCallbacks(absoluteFrameStartTime);
+      }
+
+      mCurrentIdleCallbackRunnable = null;
+    }
+
+    public void cancel() {
+      mCancelled = true;
+    }
+  }
+
   private final Object mTimerGuard = new Object();
+  private final Object mIdleCallbackGuard = new Object();
   private final PriorityQueue<Timer> mTimers;
-  private final HashMap<ExecutorToken, SparseArray<Timer>> mTimerIdsToTimers;
+  private final Map<ExecutorToken, SparseArray<Timer>> mTimerIdsToTimers;
   private final AtomicBoolean isPaused = new AtomicBoolean(true);
-  private final FrameCallback mFrameCallback = new FrameCallback();
+  private final TimerFrameCallback mTimerFrameCallback = new TimerFrameCallback();
+  private final IdleFrameCallback mIdleFrameCallback = new IdleFrameCallback();
+  private @Nullable IdleCallbackRunnable mCurrentIdleCallbackRunnable;
   private @Nullable ReactChoreographer mReactChoreographer;
   private boolean mFrameCallbackPosted = false;
+  private boolean mFrameIdleCallbackPosted = false;
+  private final Set<ExecutorToken> mSendIdleEventsExecutorTokens;
+  // Temporary array used to dipatch idle callbacks on the JS thread.
+  private final List<ExecutorToken> mIdleCallbackContextsToCall;
```

Comment:
This is not possible since we create a new runnable every frame. I don't know if there is a way to avoid creating this runnable every frame while still sending only the latest event to JS.